src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ResourceCodeDomSerializer.SerializationResourceManager.cs

```
@@ -565,7 +517,7 @@ public void SetMetadata(IDesignerSerializationManager manager, string resourceNa
                     }
 
                     // Note that when we read metadata, for backwards compatibility, we also merge in regular data from the invariant resource. We need to clear that data here, since we are going to write out metadata separately.
-                    invariant.Remove(resourceName);
+                    invariant!.Remove(resourceName);
```

How do we know this is not nullable?

---

`GetResourceSet` returns null only in an error scenario. Which I know is not satisfying... As often in internal classes, checks are pretty inconsistent

---

Generally we are moving to add comments whenever `!` is used, but given that the codebase is so old, it can be hard to determine when exactly things are null in which case we should throw a specific exception instead of getting an NRE. For this case I think the right answer would be to throw an `InvalidOperationException`, but of course if there was more context, we should throw exception that would be most helpful.

```C#
if (invariant is null) {
    throw new InvalidOperationException();
}

invariant.Remove(resourceName);
```

---

I was too quick in my first reply. Turns out that we can perfectly explain why `invariant` can't be null. I've added a comment near the null forgiving operator (which move earlier in the code).